Reapply Cygwin: path_conv: Try to handle native symlinks more sanely

orgads opened the pull request:
For local paths, add a check if the inner path components contain native
symlinks or junctions.  Compare the incoming path with the path returned
by NtQueryInformationFile(FileNameInformation).  If they differ, there
must be at least one native symlink or junction in the path.  If so,
treat the currently evaluated file as non-existant.  This forces
path_conv::check to backtrack inner path components until we eliminated
all native symlinks or junctions and have a normalized path.

Signed-off-by: Corinna Vinschen <corinna@vinschen.de>
(cherry picked from commits 456c3a46386f38887407603b2c64b7f63a4871c5,
13fd26ecf5ca8417146d57b45aed0133435c3497 and
19d59ce75d5301ae167b421111d77615eb307aa7)

github-cygwin committed:
Reapply Cygwin: path_conv: Try to handle native symlinks more sanely

For local paths, add a check if the inner path components contain native
symlinks or junctions.  Compare the incoming path with the path returned
by NtQueryInformationFile(FileNameInformation).  If they differ, there
must be at least one native symlink or junction in the path.  If so,
treat the currently evaluated file as non-existant.  This forces
path_conv::check to backtrack inner path components until we eliminated
all native symlinks or junctions and have a normalized path.

Signed-off-by: Corinna Vinschen <corinna@vinschen.de>
(cherry picked from commits 456c3a46386f38887407603b2c64b7f63a4871c5,
13fd26ecf5ca8417146d57b45aed0133435c3497 and
19d59ce75d5301ae167b421111d77615eb307aa7)

orgads commented:
Now with the subst fix.

jeremyd2019 commented:
#41

I am concerned that the subst fix of treating subst drives as symlinks will result in automatically 'dereferencing' the subst when running native Windows processes, undoing the attempted fix for too-long paths used by some build tools (which encode the *ENTIRE PATH TO THE ROOT* in the object file name 🤦)  I will test this, as well as a simple tweak:
```diff
diff --git a/winsup/cygwin/path.cc b/winsup/cygwin/path.cc
index 4ebaf4dc6..53534b8b6 100644
--- a/winsup/cygwin/path.cc
+++ b/winsup/cygwin/path.cc
@@ -3670,8 +3670,7 @@ restart:
 		   somewhere else, thus, it's a symlink in POSIX speak. */
 		if (upath.Length == 14)	/* \??\X:\ */
 		  {
-		    fileattr &= ~FILE_ATTRIBUTE_DIRECTORY;
-		    path_flags |= PATH_SYMLINK;
+		    goto file_not_symlink;
 		  }
 		/* For final paths differing in inner path components return
 		   length as negative value.  This informs path_conv::check
```

jeremyd2019 commented:
3.2.0-6
```
$ cd /t && cmd /c cd
T:\
```

dll from this PR
```
$ cd /t && cmd /c cd
E:\MINGW-packages
```

dll from this PR with above patch applied:
```
$ cd /t && cmd /c cd
T:\
```

jeremyd2019 commented:
I don't think it would normally be bad, however, there is a known issue when building packages from MINGW-packages and/or MSYS2-packages that, if the full path is too long the package build will fail.  And "too long" is not really all that long at all.

The CI scripts here work around this by moving the checkout to paths like `C:\_`.
https://github.com/msys2/MINGW-packages/actions/runs/830099565/workflow#L51

I work around it by using subst to shorten the paths.
https://github.com/jeremyd2019/msys2-build32/actions/runs/828859516/workflow#L96

If this code transparently dereferences the subst, as it seems to do, that defeats the purpose of the subst for me, and I'd probably have to switch to moving my checkout instead.  Not the end of the world but breaking my workflows (I also do this subst trick on my local machines for the same reason).

jeremyd2019 commented:
https://cygwin.com/pipermail/cygwin/2021-May/248521.html

orgads commented:
I see your point. Please refer to [my comment](https://github.com/msys2/setup-msys2/issues/136#issuecomment-842599390).

How do you suggest to proceed?

jeremyd2019 commented:
I have sent my quick and dirty patch as a reply to the "all-or-nothing" comment.  I don't know if I can make a sufficiently compelling argument about cygwin catering to brain-dead tools versus being more theoretically correct.  But since msys2's *raison d'être* is running these build tools, it may be something that this fork should work around (such as with my patch above)

jeremyd2019 committed:
Cygwin: Handle virtual drives as non-symlinks

This avoids MAX_PATH-related problems in native tools in case the
virtual drive points to a deep directory

Signed-off-by: Corinna Vinschen <corinna@vinschen.de>

jeremyd2019 commented:
My patch was applied upstream!  https://github.com/msys2/msys2-runtime/commit/c8949d04001e3dbc03651475b6cd1c5623400835

lazka commented:
nice, backported. I guess this is OK now?

orgads commented:
Fine with me. Maybe squash the commits?

jeremyd2019 commented:
I was about to indicate that it looks OK to me now, except I noticed the fs redirection assumption within `#ifdef __i386__` and thought about the x86_64-on-arm64 case.   But I just tested it and I didn't notice any issues ls-ing system32, so :+1:

lazka commented:
> Fine with me. Maybe squash the commits?

We will clean up the history on the next cygwin release.
